partiql-grammar/src/main/antlr/org/partiql/grammar/PartiQL.g4

```diff
@@ -119,15 +119,58 @@ removeCommand
 //  We essentially use the returning clause, because we currently support this with the SqlParser.
 //  See https://github.com/partiql/partiql-lang-kotlin/issues/708
 insertCommandReturning
-    : INSERT INTO pathSimple VALUE value=expr ( AT pos=expr )? onConflict? returningClause?;
+    : INSERT INTO pathSimple VALUE value=expr ( AT pos=expr )? onConflictClause? returningClause?;
 
 insertCommand
-    : INSERT INTO pathSimple VALUE value=expr ( AT pos=expr )? onConflict?  # InsertValue
-    | INSERT INTO pathSimple value=expr                                     # InsertSimple
+    : INSERT INTO pathSimple VALUE value=expr ( AT pos=expr )? onConflictClause?  # InsertLegacy
+    // See the Grammar at https://github.com/partiql/partiql-docs/blob/main/RFCs/0011-partiql-insert.md#2-proposed-grammar-and-semantics
+    | INSERT INTO symbolPrimitive asIdent? value=expr onConflictClause?           # Insert
+    ;
+
+onConflictClause
+    : ON CONFLICT WHERE expr DO NOTHING                                           # OnConflictLegacy
+    | ON CONFLICT conflictTarget? conflictAction                                  # OnConflict
     ;
 
-onConflict
-    : ON CONFLICT WHERE expr DO NOTHING;
+/**
+    <conflict target> ::=
+        ( <index target> [, <index target>]... )
+        | ( { <primary key> | <composite primary key> } )
+        | ON CONSTRAINT <constraint name>
+*/
+conflictTarget
```

Comment:
I'm curious where this will be added to the AST.

Reply:
Re-iterating my response from another comment:

The way I see it, it'll be like the following, but still need to think more about it:

```bash
(insert target::expr values::expr conflict_action::(? conflict_action) conflict_target::(? conflict_target))
```